### DISCUSSION TITLE
Game Migration - SharpDX to Vortice

### DISCUSSION THREAD

**Discussion by h1cks** [amerkoleci/Vortice.Windows#106] [Show and tell]
Migration of my game is basically done apart from a couple of items, which I will be working on.  I still have to migrate my tools which use the same engine, but it should be a lot simpler to move to 

Edit:  Things to know when migrating from Sharpdx to Vortice.  This is only the stuff I found when migrating.

- CopySubResource in sharpdx has source and destination in different orders!
- Sharpdx Vector3 returns a 0 vector on a parallel vector in a Cross product, in Numerics, this function returns a NaN on the vector.  you should check the vector with a Dot product first.
- NAudio is helpful in establishing the wave format header, see the other post in this discussion for the code.
- Matrix3x3 in Sharpdx does not exist in Numerics, you must use the Numerics.Matrix class and ensure you set the Translation component to a 0 vector if only multiplying rotation matrices.  Yes, using Transform normal will ensure it uses the 3x3 component, but if pre multiplying your matrices, you will introduce translation into the calculation.
- Color3 and Color4 are not interchangeable with Vector3 and Vector4 in Vortice, so you may find when you errors here.

--------------------------------------------------------------------------------------------------------------------

Migration has been pretty simple, most of the time is migrating has been consumed by migration off Sharpdx's implementation of Matrix and Vector3.  This was responsible for over 2000 compile errors on its own.  im down to my last 50 (all classes resolution issues).

Goal will be to implement:

(implemented by Amer)

- [x] **Direct Input** - issue of extern reference in Dinput.h to joystickstate is causing issues.  Once I work out why this is an issue, I will then try to build it closer to the vision in Vortice.

- [ ] **Media Foundation** - At some point will look into this.
- [x] (implemented by Amer)**DataStream class** - Very helpful, will pull out of Sharpdx and create a util library out of it.

Once again, thanks for the API.  It's certainly given more life back to future gamedev in .NET.

**Comment by h1cks** [comment]
For reference - https://insanesoftware.itch.io/pacific-strafe-alpha is my WIP deployment.

**Comment by amerkoleci** [comment]
Excellent stuff, I plan to add DataStream as well, not sure when.
Media Foundation is huge and not sure if I want to pull all that stuff.
DirectInput - Could be added but still legacy stuff

**Reply by h1cks** [reply]
Yeah.  My challenge will be finding a video player to replace media foundation in directx.   The irony of direct input being legacy is that it still is really the only way to get full coverage of gaming and other input devices.

**Comment by h1cks** [comment]
Extern issue was due to the fact I was using the Dinput.h file from win sdk 2010, the latest SDK has the 2 problem lines removed from the header.  See how I go now.

**Comment by h1cks** [comment]
Generation of Propertys is creating spurious code.  unsure of the origin as still learning SharpGen.  I've used the same properties as Sharpdx initially, just so I can generate the code.  But generates inherently different code.

**Comment by h1cks** [comment]
The extern issue has been sorted, but still have issues with the rules around handling PropertyGuids.

in the dinput.H file, a property is defined as:

```
#ifdef __cplusplus
#define MAKEDIPROP(prop)    (*(const GUID *)(prop))
#else
#define MAKEDIPROP(prop)    ((REFGUID)(prop))
#endif

#define DIPROP_BUFFERSIZE       MAKEDIPROP(1)
```

in the vortice generated header file its expanded to

```
// Variable created from: const from-macro:DIPROP_(.*) class:Vortice.DirectInput.DIPropertyGuids cpp-type:int cpp-cast:(int)& type:System.IntPtr name:$1 value:new System.IntPtr($1)
extern "C" int DIPROP_BUFFERSIZE__sharpgen_var__ = (int)&DIPROP_BUFFERSIZE;
```

The final code in the Functions class is (_GUID seems to be a default name created as GUID is already defined from what I understand)

```
public static readonly System.IntPtr Buffersize = new System.IntPtr(unchecked ((int)int) & (*( const _GUID * )(1) ) );
```

looking at the original sharpdx implemetation.  It supposedly should be:

```
public static readonly System.IntPtr Buffersize = new System.IntPtr(1);
```

Driving me mad, I will update the naming once I figure this last issue out with DirectInput

**Comment by h1cks** [comment]
@amerkoleci update on progress, code is in my fork at the moment.

**Comment by amerkoleci** [comment]
DataStream just landed.

**Reply by h1cks** [reply]
awesome, I know can remove using datastream = sharpdx.datastream; from my code :)

**Comment by h1cks** [comment]
@amerkoleci im almost running the game, I pulled in Sharpdx directinput to allow for the game to run for now.  I am now fixing errors in the code, most are nothing special.  Simple mistakes.  Up to loading my assets in , hoping my 3d projection isn't broken!  Shaders are loading in from file correctly.

**Comment by h1cks** [comment]
@amerkoleci think xaudio may need some additional code around handling the headers for Waveformat.  At the moment, unless you're a specific wave format, loading in audio files is problematic.  Looking at what sharpdx did, they consumed alot of code from NAudio to parse the headers.  thanks for you work.

**Comment by h1cks** [comment]
well its rendering!

![image](https://user-images.githubusercontent.com/17155885/103437456-09dbfd80-4c7c-11eb-9748-2af0b96dfb9f.png)

![image](https://user-images.githubusercontent.com/17155885/103437619-c97d7f00-4c7d-11eb-960a-eed259fc2bda.png)

im using Sharpdx directinput at the moment, and sound isnt working.  I have some issues going to fullscreen, but expect to solve them.  Found an issue in your datastream class potentially.  Will raise an issue.

**Reply by h1cks** [reply]
Framerate is about 10% less than Sharpdx, but I have some bugs in the pipeline and expect that I will get framerate back

**Comment by amerkoleci** [comment]
Absolutely, nice job

**Reply by h1cks** [reply]
I use a lot of your library!  So far so good :)

**Comment by h1cks** [comment]
@amerkoleci just a heads up, Performance of my game in Sharpdx Vs Vortice is basically the same, in fact, the performance is identical.  At lower frame rates, Vortice seems to be slightly faster.

Compiled under .NET 5.0.  Have a good holiday

**Comment by h1cks** [comment]
Just an update:

Working on Volumetric clouds.  At the moment in my Sharpdx based editor (to be moved to Vortice!).  Lighting on clouds will be done soon.

Will be working with 3d textures, will be able to stress this out in Vortice soon.

![image](https://user-images.githubusercontent.com/17155885/131923289-a3235b38-7933-4f8b-903c-989304dc8210.png)

**Reply by amerkoleci** [reply]
Excelent stuff, looks good as well :)

**Comment by h1cks** [comment]
https://www.youtube.com/watch?v=Gx1mhxXieXQ&ab_channel=AndrewH

Update, clouds in game now

**Reply by amerkoleci** [reply]
Nice stuff!

**Reply by h1cks** [reply]
need to migrate, I cant keep sitting on sharpdx.  FMV is the only piece I need to complete, hence the sample for FMV i will write.  And thanks, vortice is a life boat to those who still use DX directly.

**Reply by amerkoleci** [reply]
I wish I had more time to work on samples as well, specially the repo here: https://github.com/amerkoleci/Vortice.Windows.Samples
But as you know its spare time project and my daily job is still taking most of my time.

**Reply by h1cks** [reply]
Same.  Work and project takes most time

**Comment by amerkoleci** [comment]
@h1cks question: what do you think about C# for gamedev? How are the performance? What are the pros and cons from your perspective and experience?

**Reply by h1cks** [reply]
The performance I feel is quite good, there are a few things I avoid when developing.

* avoid pressure on the garbage collector and stack.  For my vertices, I actually use classes rather than structs, besides the fact that even if you use ref/in/out there are still some severe cases of boxing/unboxing.  
* Pre-allocating on your templates lists and dictionaries.
* If were to do things again,  I may use Span more on contiguous memory, but at the end of the day it also makes things harder if you try to pack everything.
* Again, I can't stress how much memory handling plays into performance.  
* Threading is awesome of c#, I use it alot.
* c++ is at least 3 times faster, but in this day and age I havent written a game that needs that much more performance.  Would be nice though.
* C# allowed me to develop my code faster, it just had more easily useable libraries.  C++ and linking.,...ugh.

I will add more as I think of it.  But I have been pretty happy with dev in c#

**Reply by amerkoleci** [reply]
How you measure that C++ is 3 times faster?

**Reply by h1cks** [reply]
Looking at most of the benchmark tests I've read.     In most cases and pure c++ code runs significantly faster.   But this doesn't translate if we are talking graphics as the only really difference in speed is preparation code to call direct x.  After that its the same performance.

**Reply by amerkoleci** [reply]
Well if you find any directx11 calls that cause GC allocation open issues so we can track those, native calls are done using calli and should be fast.

**Reply by h1cks** [reply]
No issues with the Vortice, none.  It's great, and the reason why I stick to c# game dev.   

This is just down 2 speed between c# which is still very fast and c++ (in respects to code gen, not Vortice's interface).  I'm actually taking aim at my code in my game.

Benchmarks like this -> https://benchmarksgame-team.pages.debian.net/benchmarksgame/fastest/csharpcore-gpp.html 

C# is incredibly competitive in a lot of aspects.

**Comment by h1cks** [comment]
My 3 times comment is in respects to code timings on pure c++ code.  The wrapper would be causing minimal performance overhead if none.

**Comment by amerkoleci** [comment]
An example of DirectInput can be seen here: https://github.com/roy-t/MiniEngine3/tree/main/src/Backend/Mini.Engine.Input

**Reply by h1cks** [reply]
@amerkoleci cheers, I have my input working now.  I now only have to do FMV!  Although, migration will be harder as I have to remigrate with so much code added since I last moved!

Current look of game.

![image](https://user-images.githubusercontent.com/17155885/137406311-37e99107-8ed4-4391-99ef-bd3ca8db8b2f.png)

**Comment by lellid** [comment]
Hi, I used SharpDX with DirectX 10. Now I'm trying to convert to Vortice with DirectX 11, but I'm missing the effects framework, i.e. classes like `ID3DX11Effect` or `ID3DX11EffectTechnique`. Is this implemented, and if yes, where can I find it?

**Reply by amerkoleci** [reply]
Nope and won't probably be added, i don't plan to add the legacy and deprecated fx framework, sorry! :)

**Comment by h1cks** [comment]
@amerkoleci all working again!  Some other changes I have to incorporate.  But its feature parity with Sharpdx except for media player.

Clouds working...found the last issue.

![image](https://user-images.githubusercontent.com/17155885/141259330-690bc851-f765-4eaa-8363-7589e3843872.png)

**Comment by h1cks** [comment]
In terms of DX11,  I'm using about 80% of the features and it's showing no issues.  Memory seems to be equivalent.  The best news is that frame rates are basically identicle because I am GPU limited (well, 650 FPS is fine! at my test harness).
